Selftest to unittest

levitte opened the pull request:
##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change

We have a few self tests a little here and there.  Having them like that leads to bit rot, and they should really be performed as part of our tests.  This is an effort to convert them into proper unit tests.

We do this by linking the program with the specific object files (through their C source files in `build.info`) that are tested.

levitte committed:
Allow indented comments in build.info

levitte committed:
Convert poly1305 selftest into internal test

levitte committed:
VMS: ignore multiply defined symbols when linking programs

The Unix and Windows linkers appear to simply ignore if any symbol is
defined multiple times in different object files and libraries.

The VMS linker, on the other hand, warns about it, loud and clear.  It
will still create the executable, but does so screaming.  So we
complicate things by saving the linker output, look through all the
errors and warnings, and if they are only made up of %LINK-W-MULDEF,
we let it pass, otherwise we output the linker output and raise the
same exit code we got from the linker.

levitte committed:
Convert asn1 selftests (a_strnid and ameth_lib) into internal test

ekasper commented on test/unit_asn1_test.c:
```diff
@@ -0,0 +1,107 @@
+
+/**********************************************************************
+ *
+ * Test of keccak1600
```
Fluke, these are ASN.1 tests.

levitte replied:
Oops.  Will fix

ekasper commented on test/unit_poly1305_test.c:
```diff
@@ -0,0 +1,552 @@
+{
+    int i = OPENSSL_hexchar2int(h);
+
+    if (i < 0)
```
We have `TEST_check(i >= 0)` for this purpose :)

levitte replied:
Ah, cool.  Will do

levitte replied:
Done

ekasper commented on test/unit_poly1305_test.c:
```diff
@@ -0,0 +1,552 @@
+    return i;
+}
+
+static void hex_decode(unsigned char *out, const char *hex)
```
Almost every test needs hex helpers so I think these should be cleaned up and moved to a common header (`testutil.h`?). However, I know you're just moving code around and I don't want to make you do the extra work so feel free to stick a `TODO(emilia)` in there. There's a bunch of duplication here that I plan to work on, anyhow.

levitte replied:
I'll stick a `TODO` for now.

ekasper commented on test/unit_poly1305_test.c:
```diff
@@ -0,0 +1,552 @@
+    result = run_tests(argv[0]);
+
+# ifdef OPENSSL_CPUID_OBJ
+    if (verbose) {
```
Verbose probably isn't the right flag here - `-benchmark`?

Also, I'd make this `#error` if benchmarking was requested but `OPENSSL_CPUID_OBJ` wasn't defined. Also, move the benchmark to a subroutine so it's easier to see what the main does?

levitte replied:
Good ideas all around.

ekasper commented on test/unit_poly1305_test.c:
```diff
@@ -0,0 +1,552 @@
+    hex_decode(key, test->keyhex);
+    hex_decode(expected, test->outhex);
+
+    in = malloc(inlen);
```
Could you change these to `OPENSSL_malloc` / `OPENSSL_free`?

levitte replied:
Sure

levitte replied:
Done

ekasper commented on test/unit_asn1_test.c:
```diff
@@ -0,0 +1,107 @@
+int main(int argc, char **argv)
+{
+    if (argc == 2 && strcmp(argv[1], "-v") == 0)
+        verbose = 1;
```
I think `verbose` is unused in this test?

levitte replied:
Removed

ekasper commented on test/unit_poly1305_test.c:
```diff
@@ -0,0 +1,552 @@
+/*
+ * Copyright 1999-2016 The OpenSSL Project Authors. All Rights Reserved.
```
Should this say simply `Copyright 2016` on new files?

levitte replied:
Yes, it shouldd

ekasper reviewed (changes requested):
This looks really neat!

I tried to not focus on the old code too much because I know you only moved it around, and I don't want to ask you to fix it all. But I left a few drive-by comments, hope that's ok.

Also, you could try running indent on the new files - there are style nits in the copied-over code that I didn't want to bother you with, but perhaps indent will take care of them.

levitte commented on crypto/asn1/tbl_standard.h:
```diff
@@ -0,0 +1,61 @@
+ */
+
+/*
+ * Now the tables and helper functions for the string table:
```
Good point

levitte replied:
Done

levitte commented on test/unit_asn1_test.c:
```diff
@@ -0,0 +1,107 @@
+ */
+
+/* Unit tests for the asn1 module */
+/* This program must be explicitelly linked with the object files it tests,
```
Good point, will do

levitte replied:
Done

levitte commented:
> I tried to not focus on the old code too much because I know you only moved it around, and I don't want to ask you to fix it all. But I left a few drive-by comments, hope that's ok.

That's perfectly fine.  It gives me input going forward as well, you know?

> Also, you could try running indent on the new files - there are style nits in the copied-over code that I didn't want to bother you with, but perhaps indent will take care of them.

I'm surprised...  after all, the code I copied has already been formatted, all I did was to adjust the indentation.  Ah well, I'll have a look at that when I finish up.

levitte commented on Configurations/descrip.mms.tmpl:
```diff
@@ -751,13 +751,35 @@ EOF
                              "WRITE OPT_FILE \"$x\"" } @deps)
           || "\@ !";
+      # The linking commands looks a bit complex, but it's for good reason.
+      # If there are multiply defined symbols (and for some of our unit tests,
```
Done

levitte commented on test/build.info:
```diff
@@ -281,6 +281,21 @@ IF[{- !$disabled{tests} -}]
     DEPEND[wpackettest]=../libcrypto ../libssl
   ENDIF
+
+  PROGRAMS_NO_INST=unit_poly1305_test unit_asn1_test
```
Done

levitte committed:
Explain the deal with internal test programs

ekasper commented on crypto/asn1/tbl_standard.h:
```diff
@@ -0,0 +1,61 @@
+/*
```
Change the main variable name as well? :)

Anyway, not so important. Perhaps I should bring up an addition to the style guide, though.

ekasper commented:
> I'm surprised... after all, the code I copied has already been formatted, all I did was to adjust the indentation. Ah well, I'll have a look at that when I finish up.

I think the poly1305 code was added after the autoformat? Anyway, I just noticed some odd whitespace, again, not crucial.

levitte commented:
> > I'm surprised... after all, the code I copied has already been formatted, all I did was to adjust the indentation. Ah well, I'll have a look at that when I finish up.

> I think the poly1305 code was added after the autoformat? Anyway, I just noticed some odd whitespace, again, not crucial.

Good point.

levitte commented:
@ekasper, did you cross the "tests are added or updated"?  I'll have you note that there is no test recipe yet (I will, but...), so that check is kinda dubious ;-)

ekasper commented:
@levitte - dang, I forgot about the recipe. I'll wait for it then.

levitte committed:
Convert modes selftests (cts128 and gcm128) to internal test

levitte committed:
Convert x509 selftests to internal test

levitte committed:
Convert mdc2 test print to internal test

levitte committed:
Finally, add a test recipe for the internal tests

levitte commented:
Awright, I've done the max I could for now, including added a test recipe.

levitte commented:
There are some remaining that I haven't dealt with yet, for diverse reasons

-   `crypto/evp/p5_crpt2.c`
    This is a trivial test to convert, but I'm surprised, is this not tested somewhere else, if not explicitly then implicitly?
-   `crypto/objects/obj_xref.c`
    Seriously not sure what to test here.  Nice printout, but...
-   `crypto/rc2/tab.c`
    That's not a test program, it's a hack to produce the key table in `crypto/rc2/rc2_skey.c`.  Should we keep it around?  Should we convert it to a perl script that's run by `make update`?  Should we just toss it?
-   `crypto/sha/keccak1600.c`
    Exists in master only and is work in progress, so I'm leaving it for now.  When @dot-asm is done with his SHA3 work, it may be time to make the selftest code into another internal test.
-   `crypto/x509v3/v3conf.c` and `crypto/x509v3/v3prin.c`
    Those are not internal tests but small utilities that hack a PEM file...  or well, they do test that perticular functionality.  However, they haven't been updated for the changed API, so we're talking bit rot big time.  @snhenson might be able to shed some light on these.  What to do with them, toss or give them some effort?

levitte committed:
Add a HEADER_MODES_H guard in include/openssl/modes.h

richsalz commented:
My view: leave the keccak stuff, obviously, but remove the others completely.

levitte commented:
So, as everyone can see with the Travis builds, there are two problems.  One (too long strings) is easy to deal with, the other not. Mac OS X has a linker that doesn't like multiply defined symbols and complains loudly...  anyone know how to deal with that?

ekasper commented:
Ohhh... so I was confused why you're getting mulitply defined symbols in the first place since the local symbols shouldn't be exported... but from what I can see, it's only because there is no sane way to pull in `OPENSSL_rdtsc` (Why isn't this declared in a header??) So, a pragmatic option would be to remove the benchmarks, I think that might be enough?

ekasper commented:
Re: the undealt ones, I agree with @richsalz: just remove them (apart from the keccak WIP). They're most likely completely dead weight but if anyone needs them there's git history and they're welcome to convert them into tests or scripts. Plus we already agreed to drop printf-debugging, and those seem to fall under that category mostly.

levitte commented:
> Ohhh... so I was confused why you're getting mulitply defined symbols in the first place since the local symbols shouldn't be exported... but from what I can see, it's only because there is no sane way to pull in `OPENSSL_rdtsc` (Why isn't this declared in a header??) So, a pragmatic option would be to remove the benchmarks, I think that might be enough?

Well, `modes_internal_test` only clashes with those select symbols because it pulls in very little else.  Have a look at everything `x509_internal_test` and imagine the chaos.

*However*, I just noticed that this only happens on `no-shared` configs, i.e. we're dealing with a static library, which is not much more than a collection of object files.  Shared libraries are dealt with a bit differently, obviously.  Sooooo, one answer could be to only enable internal tests when building a `shared` configuration, at least on MacOS X (or shall we do that generally?).

levitte commented:
> Re: the undealt ones, I agree with @richsalz: just remove them (apart from the keccak WIP). They're most likely completely dead weight but if anyone needs them there's git history and they're welcome to convert them into tests or scripts. Plus we already agreed to drop printf-debugging, and those seem to fall under that category mostly.

I generally agree.  The one that has worried me is that one in `obj_xref.c`, but seeing that code has very obviously not been updated for ages (it fails to compile as it is, the structures it looks at doesn't even look the same any more), it would seems it has passed it's usefulness...

ekasper commented:
Umm... The `.o` should be able to `#include` libcrypto headers and link against it without problems. It is, as far as the build is concerned, an independent program. It just shouldn't redefine the symbols that libcrypto exports.

IIUC the clash is because you're pulling in the _sources_ (`cryptlib.c` as well as some asm) directly (rather than the headers) - so now both your `.o` and the library define the same symbols. Or am I misunderstanding it completely?

levitte commented:
> *However*, I just noticed that this only happens on `no-shared` configs, i.e. we're dealing with a static library, which is not much more than a collection of object files.  Shared libraries are dealt with a bit differently, obviously.  Sooooo, one answer could be to only enable internal tests when building a `shared` configuration, at least on MacOS X (or shall we do that generally?).

Or dang, I'm so daft!  When linking with the static library, there isn't a need to pull in all those other files!  That is, of course, the way to solve this issue!

levitte commented:
> Umm... The `.o` should be able to `#include` libcrypto headers and link against it without problems. It is, as far as the build is concerned, an independent program. It just shouldn't redefine the symbols that libcrypto exports.

It needs to use those files directly when linking with shared libraries.  *Those symbols aren't available otherwise*

> IIUC the clash is because you're pulling in the _sources_ (`cryptlib.c` as well as some asm) directly (rather than the headers) - so now both your `.o` and the library define the same symbols. Or am I misunderstanding it completely?

You misunderstand the `build.info` files.  a `SOURCE[foo]=foo.c` creates the following dependencies:

- `foo` -> `foo.o`
- `foo.o` -> `foo.c`

However, since the final build file will never contain the same target twice, this:

```
PROGRAMS=foo
SOURCE[foo]=foo.c common.c

PROGRAMS_NO_INST=foo_internal
SOURCE[foo_internal]=foo_internal.c common.c
```

Will produce something like this:

``` make
foo : foo.o common.o
	cc -o foo foo.o common.o
foo_internal : foo_internal.o common.o
	cc -o foo_internal foo_internal.o common.o
foo.o : foo.c
	cc -c -o foo.o foo.c
foo_internal.o : foo_internal.c
	cc -c -o foo_internal.o foo_internal.c
common.o : common.c
	cc -c -o common.o common.c
```

So referring to a source file is essentially the same as referring to an object file.

ekasper commented:
> It needs to use those files directly when linking with shared libraries. Those symbols aren't available otherwise

But in non-shared builds, they're now available twice. I think we're in violent agreement about the problem now.

I hope this exercise serves as inspiration for layering the build :)

levitte commented:
> But in non-shared builds, they're now available twice.

See my comment on being daft higher up ;-)

> I think we're in violent agreement about the problem now.

Yes.  And furthermore, Mac OS X is happy now :-)

Now, to fix those strings in `poly1305_internal_test.c`

kaduk commented:
I guess I'm a little confused about why we're caving in to the "too-long strings" warning -- do we know of any platforms that we support or might support that do not support longer strings?  I am told that BoringSSL hasn't had any issues with the long string literals.
(If we do end up keeping string literal lengths down, #1775 has a glue function to do so.)

richsalz commented:
We conly require ANSI, which only requires 509 chars

levitte commented:
I made them byte arrays instead.

levitte commented:
And we seem to have success :-)

levitte committed:
Clean away remaining 'selftest' code

All of these don't compile cleanly any more, probably haven't for quite
some time

levitte commented:
Awright, did a bit of commit cleanup.  Taking this out of WIP.

ekasper commented on test/modes_internal_test.c:
```diff
@@ -0,0 +1,1077 @@
+static void benchmark_gcm128(const unsigned char *K, size_t Klen,
+                             const unsigned char *IV, size_t IVlen)
+{
+#ifdef OPENSSL_CPUID_OBJ
```
Also `#else` print an error here, like in the other benchmarks?

levitte replied:
Done

ekasper commented on test/x509_internal_test.c:
```diff
@@ -0,0 +1,102 @@
+        for (i = 0; i < STANDARD_EXTENSION_COUNT; i++, tmp++)
+            printf("%d : %s\n", (*tmp)->ext_nid, OBJ_nid2sn((*tmp)->ext_nid));
+    } else
+        fprintf(stderr, "Order OK\n");
```
It's really not clear to me what the distinction is between printing to `stderr` and `stdout`... Does anyone ever use the distinction to filter one of the two out? Choose something consistent, for example
1) Always print to stderr.
2) Test driver (`testutil.h`) prints to stdout and tests spit additional info to stderr.

Also note that switching between stdout and stderr in the test causes the Perl harness to spit out test output out of order.

levitte replied:
Anyway, I changed all `print(` to `fprintf(stderr, ` in the `*_internal_test.c`, except for the benchmark output and the help output.

levitte commented on crypto/asn1/standard_methods.h:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2006-2016 The OpenSSL Project Authors. All Rights Reserved.
```
I kinda disagree.  The table is copied verbatim from `crypto/asn1/ameth_lib.c`, which has that (c) range, and since the (c) applies on text, not on file name, it should follow.

levitte commented on test/asn1_internal_test.c:
```diff
@@ -0,0 +1,151 @@
+ *
+ ***/
+
+static SIMPLE_FIXTURE setup_tbl_std(const char *const test_case_name)
```
I'll meet you half way and rename `tbl_std` -> `tbl_standard`, since the latter is the actual variable name.

levitte replied:
Did the same with `stdmethods`, btw

levitte commented on test/build.info:
```diff
@@ -281,6 +281,81 @@ IF[{- !$disabled{tests} -}]
     DEPEND[wpackettest]=../libcrypto ../libssl
   ENDIF
+
+  # Internal test programs.  This are essentially a collection of selftest
```
I think 'selftest' -> 'internal test'.  Do you disagree?

levitte commented on test/build.info:
```diff
@@ -281,6 +281,81 @@ IF[{- !$disabled{tests} -}]
+  # object files will be chosen before those in the libraries.  This is handled
+  # properly by all linkers.
+  # Note that when building with static libraries, none of those extra files
+  # are needed, since all symbols are available anyway, ragardless of what's
```
Done

levitte commented on Configurations/descrip.mms.tmpl:
```diff
@@ -751,13 +751,46 @@ EOF
           || "\@ !";
+      # The linking commands looks a bit complex, but it's for good reason.
+      # When you link, say, foo.obj, bar.obj and libsomething.exe/share, and
+      # bar.obj happens to have a simbol that also exists in libsomething.exe,
```
Done

levitte commented on Configurations/descrip.mms.tmpl:
```diff
@@ -751,13 +751,46 @@ EOF
+      # On Unix and on Windows, the corresponding maneuvre goes through
+      # silently with the same effect.
+      # With some test programs, made for checking the internals of OpenSSL,
+      # we do this kind of linking deliberatly, picking a few specific object
```
Done

levitte commented on test/mdc2_internal_test.c:
```diff
@@ -0,0 +1,95 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Unit tests for the mdc2 module */
```
It isn't disabled by default.  I thought it was and was surprised to discover I was wrong.

levitte replied:
md2 is, however, disabled by default.  I'm guessing that's where we both get mixed up...

levitte commented on test/modes_internal_test.c:
```diff
@@ -0,0 +1,1077 @@
+    CTS128_TEST_VECTOR(64),
+};
+
+static AES_KEY *cts128_encks()
```
Actually, `ks` = `key_schedule`.

Done

levitte commented on test/x509_internal_test.c:
```diff
@@ -0,0 +1,102 @@
+        fprintf(stderr, "Extensions out of order!\n");
+        for (i = 0; i < STANDARD_EXTENSION_COUNT; i++, tmp++)
+            printf("%d : %s\n", (*tmp)->ext_nid, OBJ_nid2sn((*tmp)->ext_nid));
+    } else
```
Done

levitte commented on test/x509_internal_test.c:
```diff
@@ -0,0 +1,102 @@
+int main(int argc, char **argv)
+{
+    if (argc == 2 && strcmp(argv[1], "-v") == 0)
+        verbose = 1;
```
Removed

levitte commented on test/recipes/03-test_internal.t:
```diff
@@ -0,0 +1,33 @@
+  ( asn1_internal_test => 1,
+    modes_internal_test => 1,
+    mdc2_internal_test => 1,
+    poly1305_internal_test => 1,
```
Yup.  Done

levitte commented:
Fixups according to latest comments pushed

richsalz commented:
https://github.com/openssl/openssl/pull/1814

FdaSilvaYY commented on test/build.info:
```diff
@@ -282,7 +282,7 @@ IF[{- !$disabled{tests} -}]
   ENDIF
 
-  # Internal test programs.  This are essentially a collection of selftest
+  # Internal test programs.  Thes are essentially a collection of internal test
```
"Thes " typo missed.

levitte replied:
Fixed

levitte commented:
Did some more cleaning up and fixed the last comment.

levitte commented:
... and made compiling and running mdc2 and poly1305 internal tests depend on if they are disabled or not.

ekasper commented on crypto/x509v3/standard_exts.h:
```diff
@@ -0,0 +1,77 @@
+
+/* Number of standard extensions */
+
+#define STANDARD_EXTENSION_COUNT OSSL_NELEM(standard_exts)
```
Ah, sorry, didn't realise this was carry-over from old code.

levitte committed:
[to be squashed] test recipe fixups

ekasper commented:
I think we've hit the limit of github review usability - I'm having trouble figuring out what's changed since the last review. But I think this is good to go, thanks for the massive cleanup!

ekasper commented:
Btw coverage increased by 0.2%, woot!

levitte commented:
> I'm having trouble figuring out what's changed since the last review.

Considering I cleaned up the commits a couple of times, I can understand why...

levitte commented:
Merged!  Thanks for a great review
